File: custom_components/lucarne_family/recurrence.py
"""Recurrence engine for the Lucarne Family integration.

Thin wrapper around dateutil.rrule for RRULE parsing and date math.
All callers must use these functions — never hand-roll date calculations.

Phasing note for INTERVAL>1 rules (e.g. every 6 months, every other week):
  All rules are anchored to 1970-01-01 (_EPOCH). This means the phase of any
  interval rule is fixed globally rather than per-task. Examples:
    - FREQ=MONTHLY;BYMONTHDAY=15;INTERVAL=6  fires on Jan 15 / Jul 15 every year
    - FREQ=WEEKLY;BYDAY=MO;INTERVAL=2        fires on bi-weekly Mondays from epoch
  Phase 3 callers that need per-task phasing (e.g. "next due 6 months after
  last completion") must bump the task's `due` date directly and NOT rely on
  next_due() to compute a user-relative offset.
"""
from __future__ import annotations

import re
from collections.abc import Callable
from datetime import UTC, date, datetime
from typing import TYPE_CHECKING

from dateutil.rrule import rrule, rrulestr

if TYPE_CHECKING:

    from datetime import tzinfo

    from homeassistant.core import HomeAssistant

    from .store import LucarneFamilyStore

# Allowed RRULE patterns (see phase-2-task-model.md Supported RRULE pattern contract).
# Anything outside this set is rejected by the add_task service.
_WD = r"(?:MO|TU|WE|TH|FR|SA|SU)"
_POS_INTERVAL = r"[1-9]\d*"  # positive integer, no zero

_DAILY_RE = re.compile(rf"^FREQ=DAILY(?:;INTERVAL={_POS_INTERVAL})?$")
_WEEKLY_RE = re.compile(
    rf"^FREQ=WEEKLY;BYDAY=(?:{_WD})(?:,(?:{_WD}))*(?:;INTERVAL={_POS_INTERVAL})?$"
)
_MONTHLY_DATE_RE = re.compile(
    rf"^FREQ=MONTHLY;BYMONTHDAY=(?:[1-9]|[12]\d|3[01])(?:;INTERVAL={_POS_INTERVAL})?$"
)
_MONTHLY_NTH_RE = re.compile(
    rf"^FREQ=MONTHLY;BYDAY=[+-]?[1-5](?:{_WD[3:-1]})(?:;INTERVAL={_POS_INTERVAL})?$"
)
_YEARLY_RE = re.compile(
    rf"^FREQ=YEARLY;BYMONTH=(?:[1-9]|1[0-2]);BYMONTHDAY=(?:[1-9]|[12]\d|3[01])"
    rf"(?:;INTERVAL={_POS_INTERVAL})?$"
)

_ALLOWED_PATTERNS = (
    _DAILY_RE,
    _WEEKLY_RE,
    _MONTHLY_DATE_RE,
    _MONTHLY_NTH_RE,
    _YEARLY_RE,
)

# Weekday abbrev → human-readable
_WEEKDAY_NAMES = {
    "MO": "Monday",
    "TU": "Tuesday",
    "WE": "Wednesday",
    "TH": "Thursday",
    "FR": "Friday",
    "SA": "Saturday",
    "SU": "Sunday",
}

_MONTH_NAMES = {
    1: "January", 2: "February", 3: "March", 4: "April",
    5: "May", 6: "June", 7: "July", 8: "August",
    9: "September", 10: "October", 11: "November", 12: "December",
}

_ORDINAL = {
    1: "1st", 2: "2nd", 3: "3rd",
    -1: "Last", -2: "2nd-to-last", -3: "3rd-to-last", -4: "4th-to-last", -5: "5th-to-last",
}


def is_valid_rrule(rrule_str: str) -> bool:
    """Return True if the RRULE string is in the allowed set (or empty)."""
    if not rrule_str:
        return True
    return any(p.match(rrule_str) for p in _ALLOWED_PATTERNS)


_EPOCH = datetime(1970, 1, 1, 0, 0, 0)


def parse(rrule_str: str) -> rrule | None:
    """Parse an RRULE string; return None for empty string.

    Always anchors dtstart to 1970-01-01 so callers can query any date range.
    """
    if not rrule_str:
        return None
    return rrulestr(rrule_str, ignoretz=True, dtstart=_EPOCH)  # type: ignore[return-value]


def next_due(rrule_str: str, after: datetime) -> datetime | None:
    """Return the first occurrence strictly after `after`, or None.

    `after` may be timezone-aware; it is stripped to naive before comparison
    because our rules are anchored at naive _EPOCH.
    """
    rule = parse(rrule_str)
    if rule is None:
        return None
    naive_after = after.replace(tzinfo=None) if after.tzinfo is not None else after
    result = rule.after(naive_after, inc=False)
    return result  # type: ignore[return-value]


def is_due_today(rrule_str: str, today: date, _tz: tzinfo) -> bool:
    """Return True if the RRULE has any occurrence on `today`.

    `_tz` is accepted for API consistency with callers operating in a local
    timezone. Our RRULEs are date-only (no BYHOUR/BYMINUTE), so dateutil runs
    with naive datetimes and the day boundary is exact regardless of timezone.
    """
    rule = parse(rrule_str)
    if rule is None:
        return False
    start = datetime(today.year, today.month, today.day, 0, 0, 0)
    occ = rule.between(start, datetime(today.year, today.month, today.day, 23, 59, 59), inc=True)
    return len(occ) > 0


def friendly_summary(rrule_str: str) -> str:
    """Return a human-readable description of the RRULE."""
    if not rrule_str:
        return "One-off"

    if _DAILY_RE.match(rrule_str):
        interval = _extract_interval(rrule_str)
        if interval == 1:
            return "Daily"
        return f"Every {interval} days"

    if _WEEKLY_RE.match(rrule_str):
        interval = _extract_interval(rrule_str)
        days_part = re.search(r"BYDAY=([A-Z,]+)", rrule_str)
        day_codes: list[str] = []
        days_str = ""
        if days_part:
            day_codes = days_part.group(1).split(",")
            day_names = [_WEEKDAY_NAMES.get(d, d) for d in day_codes]
            days_str = ", ".join(day_names)
        if interval == 1:
            return f"Every {days_str}"
        return f"Every {interval} weeks on {days_str}"

    if _MONTHLY_DATE_RE.match(rrule_str):
        interval = _extract_interval(rrule_str)
        day_match = re.search(r"BYMONTHDAY=(\d+)", rrule_str)
        day_num = int(day_match.group(1)) if day_match else 1
        ordinal = _ordinal_suffix(day_num)
        if interval == 1:
            return f"Monthly on the {ordinal}"
        return f"Every {interval} months on the {ordinal}"

    if _MONTHLY_NTH_RE.match(rrule_str):
        interval = _extract_interval(rrule_str)
        byday_match = re.search(r"BYDAY=([+-]?\d)([A-Z]{2})", rrule_str)
        if byday_match:
            n = int(byday_match.group(1))
            day = _WEEKDAY_NAMES.get(byday_match.group(2), byday_match.group(2))
            ord_str = _ORDINAL.get(n, f"{n}th")
            if interval == 1:
                return f"Monthly on the {ord_str} {day}"
            return f"Every {interval} months on the {ord_str} {day}"

    if _YEARLY_RE.match(rrule_str):
        interval = _extract_interval(rrule_str)
        month_match = re.search(r"BYMONTH=(\d{1,2})", rrule_str)
        day_match = re.search(r"BYMONTHDAY=(\d{1,2})", rrule_str)
        month_name = _MONTH_NAMES.get(int(month_match.group(1)), "") if month_match else ""
        day_num = int(day_match.group(1)) if day_match else 1
        if interval == 1:
            return f"Yearly on {month_name} {day_num}"
        return f"Every {interval} years on {month_name} {day_num}"

    return rrule_str  # fallback for unexpected patterns


def _extract_interval(rrule_str: str) -> int:
    m = re.search(r"INTERVAL=(\d+)", rrule_str)
    return int(m.group(1)) if m else 1


def _ordinal_suffix(n: int) -> str:
    if 11 <= (n % 100) <= 13:
        return f"{n}th"
    suffixes = {1: "1st", 2: "2nd", 3: "3rd"}
    return suffixes.get(n % 10, f"{n}th")


def make_recurrence_evaluator(
    hass: HomeAssistant,
    store: LucarneFamilyStore,
    member_slug: str,
) -> Callable[[date], list[str]]:
    """Return a callable that lists routine task UIDs due on a given date.

    The returned function is synchronous (called from executor context by
    async_get_streak). Metadata and timezone are resolved once at first call
    and cached for the lifetime of the evaluator (one streak walk).
    """
    import zoneinfo as _zoneinfo

    try:
        local_tz: tzinfo = _zoneinfo.ZoneInfo(str(hass.config.time_zone))
    except Exception:
        local_tz = UTC

    _cache: list = []
    _loaded = False

    def evaluator(day: date) -> list[str]:
        nonlocal _cache, _loaded
        if not _loaded:
            _cache = store.get_task_metadata_sync(member_slug)
            _loaded = True
        return [
            t["item_uid"]
            for t in _cache
            if t.get("type") == "routine" and is_due_today(t.get("recurrence", ""), day, local_tz)
        ]

    return evaluator

File: src/shared/recurrence.ts
/**
 * Hand-rolled RRULE builder/parser for the six contract patterns locked in
 * Phase 2 Sub-Phase A. Does NOT depend on rrule.js.
 *
 * Contract table (must round-trip with recurrence.py):
 *  none      | "" (empty string)
 *  daily     | FREQ=DAILY[;INTERVAL=N]
 *  weekly    | FREQ=WEEKLY;BYDAY=<MO,TU,...>[;INTERVAL=N]
 *  monthly-d | FREQ=MONTHLY;BYMONTHDAY=<1-31>[;INTERVAL=N]
 *  monthly-n | FREQ=MONTHLY;BYDAY=<+/-N><DAY>[;INTERVAL=N]
 *  yearly    | FREQ=YEARLY;BYMONTH=<1-12>;BYMONTHDAY=<1-31>[;INTERVAL=N]
 */

export type RecurrenceMode =
  | 'none'
  | 'daily'
  | 'weekly'
  | 'monthly-date'
  | 'monthly-nth'
  | 'yearly'
  | 'unknown';

export const WEEKDAY_CODES = ['MO', 'TU', 'WE', 'TH', 'FR', 'SA', 'SU'] as const;
export type WeekdayCode = (typeof WEEKDAY_CODES)[number];

export interface RecurrenceNone {
  mode: 'none';
}
export interface RecurrenceDaily {
  mode: 'daily';
  interval?: number;
}
export interface RecurrenceWeekly {
  mode: 'weekly';
  days: WeekdayCode[];
  interval?: number;
}
export interface RecurrenceMonthlyDate {
  mode: 'monthly-date';
  dayOfMonth: number;
  interval?: number;
}
export interface RecurrenceMonthlyNth {
  mode: 'monthly-nth';
  nth: number;
  day: WeekdayCode;
  interval?: number;
}
export interface RecurrenceYearly {
  mode: 'yearly';
  month: number;
  dayOfMonth: number;
  interval?: number;
}
export interface RecurrenceUnknown {
  mode: 'unknown';
  raw: string;
}

export type ParsedRecurrence =
  | RecurrenceNone
  | RecurrenceDaily
  | RecurrenceWeekly
  | RecurrenceMonthlyDate
  | RecurrenceMonthlyNth
  | RecurrenceYearly
  | RecurrenceUnknown;

/** Parse RRULE string into structured form. Unknown patterns return {mode:'unknown', raw}. */
export function parseRRule(rrule: string): ParsedRecurrence {
  if (!rrule || rrule.trim() === '') return { mode: 'none' };

  const parts = rrule.trim().split(';');
  const props: Record<string, string> = {};
  for (const p of parts) {
    const eq = p.indexOf('=');
    if (eq === -1) return { mode: 'unknown', raw: rrule };
    props[p.slice(0, eq)] = p.slice(eq + 1);
  }

  const freq = props['FREQ'];
  let interval: number | undefined;
  if (props['INTERVAL'] !== undefined) {
    if (!/^[1-9]\d*$/.test(props['INTERVAL'])) return { mode: 'unknown', raw: rrule };
    interval = parseInt(props['INTERVAL'], 10);
  }
  const byDay = props['BYDAY'];
  const byMonthDay = props['BYMONTHDAY'];
  const byMonth = props['BYMONTH'];

  function hasOnlyKeys(...allowed: string[]): boolean {
    const allowedSet = new Set(allowed);
    return Object.keys(props).every((k) => allowedSet.has(k));
  }

  if (freq === 'DAILY' && !byDay && !byMonthDay && !byMonth) {
    if (!hasOnlyKeys('FREQ', 'INTERVAL')) return { mode: 'unknown', raw: rrule };
    return { mode: 'daily', ...(interval ? { interval } : {}) };
  }

  if (freq === 'WEEKLY' && byDay && !byMonthDay && !byMonth) {
    if (!hasOnlyKeys('FREQ', 'BYDAY', 'INTERVAL')) return { mode: 'unknown', raw: rrule };
    const days = byDay.split(',') as WeekdayCode[];
    if (!days.every((d) => (WEEKDAY_CODES as readonly string[]).includes(d))) {
      return { mode: 'unknown', raw: rrule };
    }
    return { mode: 'weekly', days, ...(interval ? { interval } : {}) };
  }

  if (freq === 'MONTHLY' && byMonthDay && !byDay && !byMonth) {
    if (!hasOnlyKeys('FREQ', 'BYMONTHDAY', 'INTERVAL')) return { mode: 'unknown', raw: rrule };
    if (!/^([1-9]|[12]\d|3[01])$/.test(byMonthDay)) return { mode: 'unknown', raw: rrule };
    const dom = parseInt(byMonthDay, 10);
    return { mode: 'monthly-date', dayOfMonth: dom, ...(interval ? { interval } : {}) };
  }

  if (freq === 'MONTHLY' && byDay && !byMonthDay && !byMonth) {
    if (!hasOnlyKeys('FREQ', 'BYDAY', 'INTERVAL')) return { mode: 'unknown', raw: rrule };
    // e.g. BYDAY=1SA (first Saturday) or BYDAY=-1MO (last Monday)
    const m = byDay.match(/^([+-]?\d+)([A-Z]{2})$/);
    if (!m) return { mode: 'unknown', raw: rrule };
    const nth = parseInt(m[1], 10);
    // Restrict to {1,2,3,4,-1}: the set the UI dropdown exposes.
    // recurrence.py accepts [+-]?[1-5], but values outside this set cannot be represented
    // in the edit form without silently corrupting the nth value on save.
    // Values like -2..-5 and 5 are returned as 'unknown' (round-trip preserved via raw pass-through).
    if (![1, 2, 3, 4, -1].includes(nth)) return { mode: 'unknown', raw: rrule };
    const day = m[2] as WeekdayCode;
    if (!(WEEKDAY_CODES as readonly string[]).includes(day)) return { mode: 'unknown', raw: rrule };
    return { mode: 'monthly-nth', nth, day, ...(interval ? { interval } : {}) };
  }

  if (freq === 'YEARLY' && byMonth && byMonthDay && !byDay) {
    if (!hasOnlyKeys('FREQ', 'BYMONTH', 'BYMONTHDAY', 'INTERVAL')) return { mode: 'unknown', raw: rrule };
    if (!/^([1-9]|1[0-2])$/.test(byMonth)) return { mode: 'unknown', raw: rrule };
    if (!/^([1-9]|[12]\d|3[01])$/.test(byMonthDay)) return { mode: 'unknown', raw: rrule };
    const month = parseInt(byMonth, 10);
    const dom = parseInt(byMonthDay, 10);
    return { mode: 'yearly', month, dayOfMonth: dom, ...(interval ? { interval } : {}) };
  }

  return { mode: 'unknown', raw: rrule };
}

/** Build RRULE string from structured form. */
export function buildRRule(parsed: Exclude<ParsedRecurrence, RecurrenceUnknown>): string {
  if (parsed.mode === 'none') return '';

  if (parsed.mode === 'daily') {
    let s = 'FREQ=DAILY';
    if (parsed.interval && parsed.interval > 1) s += `;INTERVAL=${parsed.interval}`;
    return s;
  }

  if (parsed.mode === 'weekly') {
    let s = `FREQ=WEEKLY;BYDAY=${parsed.days.join(',')}`;
    if (parsed.interval && parsed.interval > 1) s += `;INTERVAL=${parsed.interval}`;
    return s;
  }

  if (parsed.mode === 'monthly-date') {
    let s = `FREQ=MONTHLY;BYMONTHDAY=${parsed.dayOfMonth}`;
    if (parsed.interval && parsed.interval > 1) s += `;INTERVAL=${parsed.interval}`;
    return s;
  }

  if (parsed.mode === 'monthly-nth') {
    const nth = `${parsed.nth}`;
    let s = `FREQ=MONTHLY;BYDAY=${nth}${parsed.day}`;
    if (parsed.interval && parsed.interval > 1) s += `;INTERVAL=${parsed.interval}`;
    return s;
  }

  if (parsed.mode === 'yearly') {
    let s = `FREQ=YEARLY;BYMONTH=${parsed.month};BYMONTHDAY=${parsed.dayOfMonth}`;
    if (parsed.interval && parsed.interval > 1) s += `;INTERVAL=${parsed.interval}`;
    return s;
  }

  return '';
}

/** Friendly summary for display next to the picker. */
export function friendlySummary(rrule: string): string {
  const parsed = parseRRule(rrule);
  if (parsed.mode === 'none') return 'One-off (no repeat)';
  if (parsed.mode === 'unknown') return 'Custom recurrence (not editable here)';

  const interval = 'interval' in parsed && parsed.interval ? parsed.interval : 1;

  if (parsed.mode === 'daily') {
    return interval === 1 ? 'Daily' : `Every ${interval} days`;
  }

  if (parsed.mode === 'weekly') {
    const dayNames: Record<WeekdayCode, string> = {
      MO: 'Mon', TU: 'Tue', WE: 'Wed', TH: 'Thu', FR: 'Fri', SA: 'Sat', SU: 'Sun',
    };
    const days = parsed.days.map((d) => dayNames[d]).join(', ');
    return interval === 1 ? `Weekly on ${days}` : `Every ${interval} weeks on ${days}`;
  }

  if (parsed.mode === 'monthly-date') {
    const suffix = ordinalSuffix(parsed.dayOfMonth);
    return interval === 1
      ? `Monthly on the ${parsed.dayOfMonth}${suffix}`
      : `Every ${interval} months on the ${parsed.dayOfMonth}${suffix}`;
  }

  if (parsed.mode === 'monthly-nth') {
    const nth = nthLabel(parsed.nth);
    const dayNames: Record<WeekdayCode, string> = {
      MO: 'Monday', TU: 'Tuesday', WE: 'Wednesday', TH: 'Thursday',
      FR: 'Friday', SA: 'Saturday', SU: 'Sunday',
    };
    return interval === 1
      ? `Monthly on the ${nth} ${dayNames[parsed.day]}`
      : `Every ${interval} months on the ${nth} ${dayNames[parsed.day]}`;
  }

  if (parsed.mode === 'yearly') {
    const monthNames = [
      '', 'January', 'February', 'March', 'April', 'May', 'June',
      'July', 'August', 'September', 'October', 'November', 'December',
    ];
    const suffix = ordinalSuffix(parsed.dayOfMonth);
    return interval === 1
      ? `Yearly on ${monthNames[parsed.month]} ${parsed.dayOfMonth}${suffix}`
      : `Every ${interval} years on ${monthNames[parsed.month]} ${parsed.dayOfMonth}${suffix}`;
  }

  return '';
}

function ordinalSuffix(n: number): string {
  if (n >= 11 && n <= 13) return 'th';
  switch (n % 10) {
    case 1: return 'st';
    case 2: return 'nd';
    case 3: return 'rd';
    default: return 'th';
  }
}

function nthLabel(n: number): string {
  if (n === -1) return 'last';
  if (n === 1) return '1st';
  if (n === 2) return '2nd';
  if (n === 3) return '3rd';
  return `${n}th`;
}
